Update storage sync generated client code and add tests

pahnallan opened the pull request:
# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

pahnallan committed:
Update storage sync generated client code and add tests

pahnallan committed:
Add custom policy for filtering nulls,  fix lines

anpint commented on sdk/storage/Azure.ResourceManager.Storage/tests/Helpers/StorageTestBase.cs:
```diff
@@ -11,6 +11,8 @@
 using NUnit.Framework;
 using Azure.Core;
 using Azure.ResourceManager.Models;
+using Castle.DynamicProxy;
```
Looks like you're changing a file in the storage SDK. Any particular reason? I think we need to revert the changes to this file.

pahnallan replied:
This one accidentally slipped in, reverting the change now

pahnallan replied:
Just reverted it back to normal

pahnallan committed:
Revert changes in StorageTestBase

pahnallan committed:
Update autorest.md to reflect previous changes

Was initially changed back to previous version when trying to find a command that updates it. However we found documentation that states manual updates are the correct process.

ankushbindlish2 commented on sdk/storagesync/Azure.ResourceManager.StorageSync/tests/Azure.ResourceManager.StorageSync.Tests.csproj:
```diff
@@ -2,4 +2,24 @@
+    </None>
+  </ItemGroup>
+  <ItemGroup>
+    <Folder Include="SessionRecords\CloudEndpointTests\" />
```
Why do we need include in csproj?

pahnallan replied:
I believe these were automatically added when running the initial test recording, but I removed them in the latest commit and it should be resolved now.

ankushbindlish2 commented on sdk/storagesync/Azure.ResourceManager.StorageSync/tests/CustomPolicy/StorageSyncResponseNullFilterPolicy.cs:
```diff
@@ -0,0 +1,40 @@
+            //The delete method response only contains "value":null, after filter, nothing would left
+            if (message.Request.Method != RequestMethod.Delete && message.Response.ContentStream != null)
+            {
+                StreamReader reader = new StreamReader(message.Response.ContentStream);
```
1. use var
2. use "using"

ankushbindlish2 commented on sdk/storagesync/Azure.ResourceManager.StorageSync/tests/CustomPolicy/StorageSyncResponseNullFilterPolicy.cs:
```diff
@@ -0,0 +1,40 @@
+
+namespace Azure.ResourceManager.StorageSync.Tests.CustomPolicy
+{
+    public class StorageSyncResponseNullFilterPolicy : HttpPipelineSynchronousPolicy
```
Would you please add some documentation

ankushbindlish2 commented on sdk/storagesync/Azure.ResourceManager.StorageSync/tests/StorageSyncManagementTestBase.cs:
```diff
@@ -37,5 +60,21 @@ protected async Task<ResourceGroupResource> CreateResourceGroup(SubscriptionReso
+
+        public async Task<ResourceGroupResource> CreateResourceGroupAsync()
+        {
+            string resourceGroupName = Recording.GenerateAssetName("resTrack2");
```
afs-sdk-rg

ankushbindlish2 commented on sdk/storagesync/Azure.ResourceManager.StorageSync/tests/Tests/CloudEndpointTests.cs:
```diff
@@ -0,0 +1,167 @@
+        private StorageSyncServiceResource _storageSyncServiceResource;
+        private StorageSyncGroupResource _storageSyncGroupResource;
+
+        public CloudEndpointTests(bool async) : base(async) //, RecordedTestMode.Record)
```
remove

ankushbindlish2 commented on sdk/storagesync/Azure.ResourceManager.StorageSync/tests/Tests/CloudEndpointTests.cs:
```diff
@@ -0,0 +1,167 @@
+        public async Task CreateStorageSyncResources()
+        {
+            _resourceGroup = await CreateResourceGroupAsync();
+            _storageSyncServiceName = Recording.GenerateAssetName("sss-cepcreate");
```
afs-sdk-sss-create

ankushbindlish2 commented on sdk/storagesync/Azure.ResourceManager.StorageSync/tests/Tests/CloudEndpointTests.cs:
```diff
@@ -0,0 +1,167 @@
+        public async Task DeleteStorageSyncResources()
+        {
+            var storageSyncServiseExists = (await _resourceGroup.GetStorageSyncServices().ExistsAsync(_storageSyncServiceName)).Value;
+            if (storageSyncServiseExists)
```
storageSyncServiceExists

ankushbindlish2 commented on sdk/storagesync/Azure.ResourceManager.StorageSync/tests/Tests/CloudEndpointTests.cs:
```diff
@@ -0,0 +1,167 @@
+        public async Task CloudEndpointGetTest()
+        {
+            // Create CloudEndpoint
+            CloudEndpointResource cloudEndpointResource = (await _storageSyncGroupResource.GetCloudEndpoints().CreateOrUpdateAsync(WaitUntil.Completed, _cloudEndpointName, _cloudEndpointCreateOrUpdateContent)).Value;
```
Option 1 :  Get as is, Check HasValue using Assert.  Use .Value
Option 2 : Use GetValueOrDefault()

ankushbindlish2 commented on sdk/storagesync/Azure.ResourceManager.StorageSync/tests/Tests/CloudEndpointTests.cs:
```diff
@@ -0,0 +1,167 @@
+            _syncGroupName = Recording.GenerateAssetName("sg-cepcreate");
+            _cloudEndpointName = Recording.GenerateAssetName("cepcreate");
+
+            _storageSyncServiceCreateOrUpdateContent = StorageSyncManagementTestUtilities.GetDefaultStorageSyncServiceParameters();
```
Why are they not abstracted out in common test class setup - whatever make sense

ankushbindlish2 reviewed (changes requested):
I have left few changes to incorporate

ankushbindlish2 commented:
/azp run

pahnallan committed:
Fix typos and abstract away test setup

pahnallan commented:
/azp run

ArthurMa1978 commented:
/azp run net - storagesync - mgmt

ArthurMa1978 commented on sdk/storagesync/Azure.ResourceManager.StorageSync/api/Azure.ResourceManager.StorageSync.netstandard2.0.cs:
```diff
@@ -380,6 +382,12 @@ namespace Azure.ResourceManager.StorageSync.Models
         public static bool operator !=(Azure.ResourceManager.StorageSync.Models.ChangeDetectionMode left, Azure.ResourceManager.StorageSync.Models.ChangeDetectionMode right) { throw null; }
         public override string ToString() { throw null; }
     }
+    public partial class CloudEndpointAfsShareMetadataCertificatePublicKeys
```
This name is weird, not sure if we can have a better name

pahnallan replied:
This name comes from another api (AfsShareMetadataCertificatePublicKeys API) and was mostly put in swagger for compliance. It won't be used by customers so it might not be worth it to change.

ArthurMa1978 commented on sdk/storagesync/Azure.ResourceManager.StorageSync/src/autorest.md:
```diff
@@ -8,8 +8,8 @@ azure-arm: true
 namespace: Azure.ResourceManager.StorageSync
-require: https://github.com/Azure/azure-rest-api-specs/blob/2614c5038cb384e864c1608e76be61a5ee84cb02/specification/storagesync/resource-manager/readme.md
-tag: package-2020-09-01
+require: https://github.com/Azure/azure-rest-api-specs/blob/d1eee5499dbf9281debdc90c4f4cbc7470fb8d6d/specification/storagesync/resource-manager/readme.md
```
Please update the CHANGELOG.md and give the date of when do you want to release the new package

pahnallan replied:
Sounds good, just updated the CHANGELOG.md. 

Could we get this released at the next earliest date? I slotted the release for 2022-10-21 for now but it can be changed depending on your availability for publishing the release. Thanks!

pahnallan committed:
Update CHANGELOG.md

pahnallan committed:
Fix typos in CHANGELOG.md

pahnallan committed:
Remove empty sections in CHANGELOG.md

Can't be empty otherwise pipeline will throw errors

pahnallan commented:
/azp run net - storagesync - mgmt

pahnallan committed:
Bump release date in CHANGELOG
